Haddock failures on GHC 8.10

With GHC 8.10.7, `cabal v2-haddock saw-script` fails with:

```
src/SAWScript/Crucible/LLVM/Override.hs:1725:12: error:
    parse error on input ‘-- * The bitfield is in fact a bitvector, and’
     |
1725 |            -- * The bitfield is in fact a bitvector, and
     |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

And `cabal v2-haddock heapster-saw` fails with:

```
src/Verifier/SAW/Heapster/Implication.hs:5589:34: error:
    parse error on input ‘{- ^ the variable @x@ -}’
     |
5589 |   ExprVar (LLVMPointerType w) -> {- ^ the variable @x@ -}
     |                                  ^^^^^^^^^^^^^^^^^^^^^^^^
```

We should fix this and ensure that the CI runs Haddock on `saw-script`, `heapster-saw`, and related libraries to avoid regressions like this in the future.

--- a/.github/ci.sh
+++ b/.github/ci.sh
@@ -54,6 +54,34 @@ build() {
   cabal v2-build "$@" "${pkgs[@]}"
 }
 
+haddock() {
+  # It seems that the secret sauce for getting cabal to _not_ go
+  # through building docs for every single sublibrary is to pass
+  # --disable-documentation, counterintuitive though that is.
+  #
+  # Note: there's a v2-haddock-project that runs haddock on all
+  # packages in the project, which would avoid needing to list them
+  # out. However, it doesn't support the --disable-documentation
+  # option, so it won't currently serve. (Also for some reason it
+  # currently demands --internal in place of --haddock-internal.)
+  local PACKAGES='
+    rme
+    saw-core
+    cryptol-saw-core
+    saw-core-what4
+    saw-core-sbv
+    saw-core-aig
+    saw-core-coq
+    heapster-saw
+    saw-script
+    saw-remote-api
+    crucible-mir-comp
+    crux-mir-comp
+    verif-viewer
+  '
+  cabal v2-haddock --haddock-internal --disable-documentation $PACKAGES
+}
+
 # Gather and tar up all HPC coverage files and binaries
 collect_hpc_files() {
   local MIX_FILES=$(find dist-newstyle -name "*.mix")

--- a/.github/ci.sh
+++ b/.github/ci.sh
@@ -36,7 +36,13 @@ build() {
   ghc_ver="$(ghc --numeric-version)"
   cp cabal.GHC-"$ghc_ver".config cabal.project.freeze
   cabal v2-update
-  cabal v2-configure -j --enable-tests
+  # Configure with --disable-documentation and --haddock-internal so
+  # that the haddock run later, if enabled, doesn't recompile the
+  # world by using those flags. (See haddock() below for discussion of
+  # why those flags are used.) We could do this only for builds where
+  # we're intending to do the haddock run, but it should have no
+  # effect otherwise and unconditional is simpler.
+  cabal v2-configure -j --enable-tests --disable-documentation --haddock-internal
   git status --porcelain
   if $IS_WIN; then
     pkgs=(saw crux-mir-comp)
@@ -64,6 +70,10 @@ haddock() {
   # out. However, it doesn't support the --disable-documentation
   # option, so it won't currently serve. (Also for some reason it
   # currently demands --internal in place of --haddock-internal.)
+  #
+  # We use --haddock-internal because the point of generating the
+  # haddocks for SAW (which doesn't have an external-facing library
+  # interface) is to serve as an internals reference.
   local PACKAGES='
     rme
     saw-core

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -77,6 +77,7 @@ jobs:
         os: [ubuntu-22.04]
         cabal: ["3.10.3.0"]
         ghc: ["9.4.8", "9.6.6", "9.8.2"]
+        haddock: [true]
         run-tests: [true]
         hpc: [false]
         include:
@@ -86,12 +87,14 @@ jobs:
           - os: ubuntu-20.04
             ghc: "9.4.8"
             cabal: "3.10.3.0"
+            haddock: false
             run-tests: false
             hpc: false
           # Include one job with HPC enabled
           - os: ubuntu-22.04
             ghc: "9.4.8"
             cabal: "3.10.3.0"
+            haddock: false
             run-tests: true
             hpc: true
           # Windows and macOS CI runners are more expensive than Linux runners,
@@ -100,14 +103,17 @@ jobs:
           # macOS runner (macos-14).
           - os: windows-2019
             ghc: 9.4.8
+            haddock: false
             run-tests: true
             hpc: false
           - os: macos-13
             ghc: 9.4.8
+            haddock: false
             run-tests: true
             hpc: false
           - os: macos-14
             ghc: 9.4.8
+            haddock: false
             run-tests: true
             hpc: false
     outputs:
@@ -169,11 +175,14 @@ jobs:
         env:
           ENABLE_HPC: ${{ matrix.hpc }}
 
-
       - shell: bash
         env:
           RELEASE: ${{ needs.config.outputs.release }}
         run: .github/ci.sh build_cryptol
+
+      - if: matrix.haddock == true
+        shell: bash
+        run: .github/ci.sh haddock
 
       - uses: GaloisInc/.github/actions/cabal-collect-bins@v1.1.1
         id: cabal-test-suites

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -23,7 +23,7 @@ env:
   # older .so files are cached, which can have various effects
   # (e.g. cabal complains it can't find a valid version of the "happy"
   # tool).
-  CABAL_CACHE_VERSION: 3
+  CABAL_CACHE_VERSION: 4
   SOLVER_CACHE_VERSION: 1
 
   DISABLED_TESTS: "test0000 test_FNV_a1_rev test0010_jss_cnf_exp test0039_rust test_boilerplate test_external_abc"

